linkerd/protocol/http/src/main/scala/com/twitter/finagle/buoyant/linkerd/Headers.scala

```diff
@@ -1,46 +1,232 @@
 package com.twitter.finagle.buoyant.linkerd
 
-import com.twitter.finagle.{Name, Path, Service, ServiceFactory, SimpleFilter, Stack}
+import com.twitter.finagle.{Deadline => FDeadline, Dtab => FDtab, Status => _, _}
 import com.twitter.finagle.buoyant.{Dst => BuoyantDst}
+import com.twitter.finagle.context.Contexts
 import com.twitter.finagle.http._
 import com.twitter.finagle.tracing._
-import com.twitter.util.Try
+import com.twitter.util.{Future, Return, Throw, Time, Try}
 import java.util.Base64
 
 // all of this based on com.twitter.finagle.http.Codec
 
 object Headers {
   val Prefix = "l5d-"
 
-  // TODO fold dtabs into this.
   object Ctx {
-    val Key = Prefix + "ctx"
 
     /**
-     * Get a trace id from a base64 encoded buffer.
+     * A serverside stack module that extracts contextual information
+     * from requests and configures the local
+     * `com.twitter.finagle.Context` appropriately. Currently this includes:
+     *   - Deadline
+     *   - Dtab
      *
-     * Based on com.twitter.finagle.tracing.Trace.idCtx.tryUnmarshal
+     * Note that currently, the dtabs read by this module are
+     * appeneded to that specified by the `l5d-dtab` header.  The
+     * `dtab-local` header should be considered deprecated in favor of
+     * `l5d-dtab, and will not be supported in the future.
      *
-     * The wire format is (big-endian):
-     *   ''reqId:8 parentId:8 traceId:8 flags:8''
+     * Note that trace configuration is handled by [[HttpTraceInitializer.server]].
      */
-    def get(b64: String): Try[TraceId] =
-      Try(Base64.getDecoder.decode(b64)).flatMap(TraceId.deserialize)
-
-    def get(headers: HeaderMap): Option[TraceId] =
-      for {
-        header <- headers.get(Key)
-        traceId <- get(header).toOption
-      } yield traceId
-
-    def set(headers: HeaderMap, id: TraceId): Unit = {
-      val bytes = TraceId.serialize(id)
-      val b64 = Base64.getEncoder().encodeToString(bytes)
-      val _ = headers.set(Key, b64)
+    val serverModule: Stackable[ServiceFactory[Request, Response]] =
+      new Stack.Module0[ServiceFactory[Request, Response]] {
+        val role = Stack.Role("DeadlineExtract")
```

Comment:
role and description should indicate that other context is extracted as well (eg dtab)